vo_gpu: split 'alpha' into two options

occivink opened the pull request:
Currently it covers two somewhat orthogonal concepts:
* whether to create a framebuffer with alpha component
* how to blend video with alpha information

This commit creates a new flag 'fbo-with-alpha' which now controls
the first part, while 'alpha' now only covers the second.

Previously it would only be possible to make the window transparent
with 'alpha=yes', which prevented blending of the image with tiles or
the background.
This split allow using different blending strategies while keeping the
window transparent.

One downside is that only specifying 'alpha=yes' no longer makes sense
without specifying 'fbo-with-alpha' as well.

Fixes #9615
If this seems ok conceptually, I can adjust the docs and document the new flag.
Tested on wayland with wayfire (wlroots-based) and weston.

occivink commented:
While working on this I noticed a strange behavior when using framebuffers with alpha channel. When loading this image
![transparent-circles](https://user-images.githubusercontent.com/5473453/146798373-fc4c0db9-ea37-418f-b161-fd79632df47b.png)
with: `mpv --no-config --alpha=yes transparent-circles.png` (on master), it gets rendered like this
![weird](https://user-images.githubusercontent.com/5473453/146798855-2d1c4b7d-c777-42e0-86fc-9ce238501451.png)
resulting in noticeable 'rings'. These rings are not visible when the background is fully opaque (for example, with `blend=tiles`)

haasn commented:
I think I would prefer naming this option something like "--window-transparent", to go with the other "--window-" flags.

occivink commented:
Updated the option name to `window-transparent` and moved it to `options.c`. I've also noticed that toggling `window-transparent` (or previously, switching alpha to/from `yes`) at runtime doesn't work, only the initial status of the option seem to matter. I'm not sure if this is meant to be a supported usecase, but I thought I'd mention it.

Dudemanguy commented:
It's because you added the option value to `mp_vo_opts` instead of `gl_video_opts`.

occivink commented:
Ok indeed. Does the option belong there? I added it to vo_opts since that's where others window- options were defined.

Dudemanguy commented:
The `want_alpha` flag is only ever going to work with `gpu` or `gpu-next` so I say it would belong with `gl_video_opts` and not generic `mp_vo_opts`.

occivink commented:
I put the option in `gl_video_opts`, added a warning for alpha=yes without window-transparent, and updated the documentation accordingly

occivink commented:
Added an `interface-changes.rst` entry (and also ping).

haasn commented on DOCS/interface-changes.rst:
```diff
@@ -40,6 +40,8 @@ Interface changes
     - add `dolbyvision` sub-parameter to `format` video filter
     - `--sub-visibility` no longer has any effect on secondary subtitles
+    - add `--window-transparent`, which controls whether a framebuffer with alpha
+      component is requested. `--video=yes` no longer does it automatically
```
you meant `--alpha=yes`

haasn commented on DOCS/man/options.rst:
```diff
@@ -6573,17 +6573,22 @@ them.
-        system and desktop environment.
+        Render the frame as-is, preserving its alpha component. Blending is
+        done on the mpv window by the windowing system. This only works
+        if ``--window-transparent`` is enabled.
```
Should `--alpha=yes` also automatically imply `--window-transparent=yes`? Maybe `--window-transparent=auto` should be the default, which means `yes` iff `--alpha=yes` (and the file has an alpha channel).

occivink replied:
I tried to do this, but the part "and the file has an alpha channel" is more difficult than anticipated. The GL context is created before init_video() is called, and so before we have knowledge of whether the file has an alpha channel.

Even if that were not the case, we can easily imagine a case where an image without transparency is loaded, and later another one that has transparency. So the GL context would have to be recreated when that happens. 
According to @Dudemanguy this is difficult-to-not-doable
>I would only forsee problems trying to recreate contexts like this so maybe the auto mode is just a no-go
if you make a context disappear while a vo thinks it's still up and running... probably just bad things will happen

As a simpler alternative, we could still make a `window-transparent=auto` mode which only depends on the alpha value, and would be yes if `alpha=yes`, and no otherwise? This also has the advantage of not breaking the current behavior of setting `alpha=yes` by itself.

haasn replied:
> As a simpler alternative, we could still make a `window-transparent=auto` mode which only depends on the alpha value, and would be yes if `alpha=yes`, and no otherwise? This also has the advantage of not breaking the current behavior of setting `alpha=yes` by itself.

I think I would prefer this approach.

occivink replied:
This is now the case. I've also dropped the warning message since it would only matter when the user explicitly sets `alpha=yes` and `window-transparent=no` which does not make much sense. I also was not able to test with gpu-next, but since the change is very similar to gpu, I'm somewhat confident.

haasn commented on DOCS/man/options.rst:
```diff
@@ -6599,17 +6599,25 @@ them.
+    all platforms. If alpha framebuffers are unavailable, it silently falls
+    back on a normal framebuffer. Note that if you set the ``--fbo-format``
+    option to a non-default value, a format with alpha must be specified,
+    or this won't work. Whether this really works depends on the windowing
```
The sentence about `--fbo-format` should go into the documentation of `--alpha=yes`, since it's relevant only for alpha blending *within* the pipeline, rather than display onto the screen.

occivink replied:
Like this? I'll admit that I don't full understand the workings of `fbo-format`, even after trying its different values out.

occivink commented:
ping

musicfunfan commented:
So if i understand correctly this PR will let users like me to have a video with alpha channel display properly. 
What i mean is. I have video of mine that has transparent elements. My expectation from mpv-player is to play them with transparency. For example when i have a transparent element on the video i want to see what ever is behind the player. I want the mpv window to be transparent when the video is transparent. Only on the places that the video is transparent. 

Now i can not do that on my system. 
Os archlinux with X11 and picom. 

If this PR can do that. I would like that very much.

If a way to do that already exist **please** let me know.  
 
Sorry if this comment is not relevant i just searching for a way solve my problem and i find this PR.

occivink commented:
What you're trying to achieve should already be possible, all you have to do is pass `--alpha=yes`.
This PR is only relevant if you don't want the window to be fully transparent, but instead blended with the mpv background (as specified by `--background`). 

Regarding, this PR, I've noticed that it does not work with `vo=gpu-next`. Based on my understanding of the code, something in libplacebo would have to be changed. I had a look into it, but was not able to get it to work.

musicfunfan commented:
When i try to do the ``--alpha=yes`` thing the transparent stuff just turn to black.... Not transparent. :(

occivink commented:
Ah yes, it could be that on X11 + mesa this doesn't work, see also the comments in https://github.com/mpv-player/mpv/issues/9862. Can you try on a wayland compositor? Anyway this is not really on-topic for this PR, maybe ask on IRC or create a new issue.
